pippy/IR.py

```diff
@@ -1326,54 +1338,80 @@ def annotate_split_points(mod: torch.nn.Module, spec: Dict[str, SplitPoint]):
 
 
 def pipeline(
-    mod: torch.nn.Module,
+    module: torch.nn.Module,
     num_chunks: int,
     example_args: Tuple[Any, ...],
     example_kwargs: Optional[Dict[str, Any]] = None,
     split_policy: Optional[Callable[[fx.GraphModule], fx.GraphModule]] = None,
-    args_chunk_spec: Optional[Tuple[Any, ...]] = None,
-    kwargs_chunk_spec: Optional[Dict[str, Any]] = None,
-    output_chunk_spec=None,
-    constraints: Optional[List[Constraint]] = None,
 ) -> Pipe:
     """
-    pipeline creates a new Pipe by tracing the provided module.
+    Creates a pipeline representation for the provided module.
 
-    See Pipe for more details.
+    See `Pipe` for more details.
 
     Arguments
     ---------
-    mod:
-        The module to be traced.
+    module:
+        The module to be transformed into a `Pipe`.
     num_chunks:
-        The number of chunks to split the module into.
+        The number of microbatches to be run with this pipeline.
     example_args:
-        The example arguments to be used for tracing.
+        Example positional inputs to be used with this pipeline.
     example_kwargs:
-        The example keyword arguments to be used for tracing.
+        Example keyword inputs to be used with this pipeline. (default: `None`)
     split_policy:
-        The policy to use for splitting the module.
-    args_chunk_spec:
-        Args for each chunk.
-    kwargs_chunk_spec:
-        Kwargs for each chunk.
-    output_chunk_spec:
-        Outputs for each chunk.
-    constraints:
-        Constraints to be used for tracing.
+        The policy to use for splitting the module. (default: `None`)
 
     Returns
     -------
-    The traced Pipe.
+    A pipeline representation of class `Pipe`.
     """
     return Pipe.from_tracing(
-        mod=mod,
+        mod=module,
         num_chunks=num_chunks,
         example_args=example_args,
         example_kwargs=example_kwargs,
         split_policy=split_policy,
-        args_chunk_spec=args_chunk_spec,
-        kwargs_chunk_spec=kwargs_chunk_spec,
-        output_chunk_spec=output_chunk_spec,
-        constraints=constraints,
     )
+
+
+# Context manager for setting `args_chunk_spec` during creation of Pipe
+class ArgsChunkSpec:
+    def __init__(
+        self,
+        chunk_dims: Tuple[int, ...],
+    ):
+        self.args_chunk_spec = map_aggregate(
+            chunk_dims,
+            lambda dim: TensorChunkSpec(dim),
+        )
+
+    def __enter__(self):
+        # Inject into the Pipe class
+        Pipe.args_chunk_spec = self.args_chunk_spec
+        return self.args_chunk_spec
+
+    def __exit__(self, exc_type, exc_val, traceback):
+        # Remove from the Pipe class
+        Pipe.args_chunk_spec = None
+
+
+# Context manager for setting `kwargs_chunk_spec` during creation of Pipe
+class KwargsChunkSpec:
+    def __init__(
+        self,
+        chunk_dims: Dict[str, int],
+    ):
+        self.kwargs_chunk_spec = map_aggregate(
```

Comment:
Wouldn't this apply to all kwargs in the dict? In the summary description this is used like `KwargsChunkSpec({"id": 0})` so wouldn't we want to just apply to "id" key?

Reply:
So a more general example would be like:
`KwargsChunkSpec({"id": 0, "mask": 1})`
then this helper will cut the id tensor on dim 0, and cut the mask tensor on dim 1.

So yes, the intention is to apply to all kwargs, based on their respective chunk dim.

In future, we may have to support "replicate" too -- instead of chunking.